Name: Ramandeep Singh
student_id: c0904415

# 4. Model Evaluation and Testing
Comprehensive evaluation of model performance using appropriate metrics. Testing the model on unseen data and providing a thorough analysis of results.

In [1]:
import pickle  # Import the pickle module for loading saved objects

# Load the previously saved StandardScaler model from a .pkl file
sc = pickle.load(open('scaler.pkl', 'rb'))


In [2]:
import pickle  # Import the pickle module for loading saved objects

# Load the polynomial regression model from a .pkl file
poly = pickle.load(open('poly_regression.pkl', 'rb'))

# Load the Ridge regression model from a .pkl file
ridge = pickle.load(open('ridge_model.pkl', 'rb'))

# Load the Lasso regression model from a .pkl file
lasso = pickle.load(open('lasso_model.pkl', 'rb'))

# Load the Linear Regression model from a .pkl file
lr = pickle.load(open('linear_regression_model.pkl', 'rb'))

# Load the ElasticNet model from a .pkl file
en = pickle.load(open('elasticnet_model.pkl', 'rb'))

# Load the Ridge regression model that was optimized with cross-validation
ridge_CV = pickle.load(open('cross_validation_model_ridge.pkl', 'rb'))

# Load the Lasso regression model that was optimized with cross-validation
lasso_CV = pickle.load(open('cross_validation_model_lasso.pkl', 'rb'))

# Load the ElasticNet model that was optimized with cross-validation
elasticnet_CV = pickle.load(open('cross_validation_model_elasticnet.pkl', 'rb'))

# Load the hyperparameter-tuned Ridge regression model from a .pkl file
ridge_ht = pickle.load(open('hyperparameter_tuned_ridge.pkl', 'rb'))

# Load the hyperparameter-tuned Lasso regression model from a .pkl file
lasso_ht = pickle.load(open('hyperparameter_tuned_lasso.pkl', 'rb'))

# Load the hyperparameter-tuned ElasticNet regression model from a .pkl file
elasticnet_ht = pickle.load(open('hyperparameter_tuned_elasticnet.pkl', 'rb'))


In [3]:
unknown_data = {
    'Temperature': [21.0, 26.5, 28.3, 30.0, 34.0, 37.0, 39.5, 22.0, 26.8, 31.2],
    'RH': [20, 23, 60, 85, 42, 70, 45, 35, 40, 50],
    'Ws': [20, 6, 9, 12.0, 22, 27, 14, 16.5, 15, 18],
    'Rain': [2, 5, 6, 8, 9, 13, 3.0, 12, 10, 8.0],
    'FFMC': [32, 40, 45, 58, 31.5, 92, 82, 80, 68, 66],
    'DMC': [5, 48, 50, 35, 27, 60, 29, 57, 41, 11],
    'DC': [9, 95, 205, 100, 200, 150, 165, 140, 180, 190],
    'ISI': [6, 7, 19, 13, 2.0, 16, 3.0, 9, 10, 14],
    'BUI': [62, 54, 12, 25, 6, 40, 65, 20, 58, 22],
    'FWI': [3.0, 23, 7, 8, 12, 29, 28, 21, 18, 19.0]
}


In [4]:
import pandas as pd


In [5]:
# Converting the dictionary into a DataFrame
df = pd.DataFrame(unknown_data)

# Displaying the DataFrame
print(df)

   Temperature  RH    Ws  Rain  FFMC  DMC   DC   ISI  BUI   FWI
0         21.0  20  20.0   2.0  32.0    5    9   6.0   62   3.0
1         26.5  23   6.0   5.0  40.0   48   95   7.0   54  23.0
2         28.3  60   9.0   6.0  45.0   50  205  19.0   12   7.0
3         30.0  85  12.0   8.0  58.0   35  100  13.0   25   8.0
4         34.0  42  22.0   9.0  31.5   27  200   2.0    6  12.0
5         37.0  70  27.0  13.0  92.0   60  150  16.0   40  29.0
6         39.5  45  14.0   3.0  82.0   29  165   3.0   65  28.0
7         22.0  35  16.5  12.0  80.0   57  140   9.0   20  21.0
8         26.8  40  15.0  10.0  68.0   41  180  10.0   58  18.0
9         31.2  50  18.0   8.0  66.0   11  190  14.0   22  19.0


In [6]:
import numpy as np

# Initialize a dictionary to store model predictions
model_predictions = {}

# Dictionary of loaded models with their corresponding names
models = {
    
    'Ridge Regression': ridge,
    'Lasso Regression': lasso,
    'Linear Regression': lr,
    'ElasticNet': en,
    'Ridge CV': ridge_CV,
    'Lasso CV': lasso_CV,
    'ElasticNet CV': elasticnet_CV,
    'Hyperparameter Tuned Ridge': ridge_ht,
    'Hyperparameter Tuned Lasso': lasso_ht,
    'Hyperparameter Tuned ElasticNet': elasticnet_ht,
}

# Scale the input data using the loaded scaler
scaled_data = sc.transform(df)

# Iterate over each model to make predictions
for model_name, model in models.items():
    try:
        # Make predictions on the scaled data
        predictions = model.predict(scaled_data)
        
        # Store the predictions in the dictionary with the model name as the key
        model_predictions[model_name] = predictions
        
        # Print the predictions from the current model
        print(f"Predictions from {model_name}: {predictions}")
    except Exception as e:
        # Handle any errors that occur during predictions
        print(f"Error with model {model_name}: {e}")



# Convert the dictionary of predictions into a DataFrame for easier analysis
predictions_df = pd.DataFrame(model_predictions)

# Display the DataFrame containing predictions from all models
print(predictions_df)


Predictions from Ridge Regression: [-0.87498342 -0.27763929  0.27729379  0.69463764 -0.56174098  1.64910973
  0.92670381  0.7552974   0.6479302   0.9911363 ]
Predictions from Lasso Regression: [0.56701031 0.56701031 0.56701031 0.56701031 0.56701031 0.56701031
 0.56701031 0.56701031 0.56701031 0.56701031]
Predictions from Linear Regression: [-0.11495329 -0.19915883 -0.46269056  0.48911889 -1.07969344  1.33533419
  1.30543762  0.20308972  0.7322202   0.904315  ]
Predictions from ElasticNet: [0.56701031 0.56701031 0.56701031 0.56701031 0.56701031 0.56701031
 0.56701031 0.56701031 0.56701031 0.56701031]
Predictions from Ridge CV: [-0.87498342 -0.27763929  0.27729379  0.69463764 -0.56174098  1.64910973
  0.92670381  0.7552974   0.6479302   0.9911363 ]
Predictions from Lasso CV: [0.56701031 0.56701031 0.56701031 0.56701031 0.56701031 0.56701031
 0.56701031 0.56701031 0.56701031 0.56701031]
Predictions from ElasticNet CV: [0.56701031 0.56701031 0.56701031 0.56701031 0.56701031 0.56701031
 0.5

Insights based on the prediction results you provided for various regression models:

Diverse Predictions:

The Ridge Regression and Ridge CV models produced negative predictions for some instances, indicating a potential issue with the model fitting, especially if negative values are not meaningful in your context.
In contrast, Lasso Regression, ElasticNet, and their cross-validated variants consistently returned the same positive prediction (0.56701) for all instances, suggesting they may be overly simplified or encountering convergence issues.
Linear Regression Performance:

The Linear Regression model yielded a mix of positive and negative predictions, which may indicate a more balanced model fitting compared to Ridge Regression. However, the spread of the predictions still suggests that there may be a lack of robustness in the model for the given dataset.
ElasticNet Consistency:

The ElasticNet model and its hyperparameter-tuned counterpart returned the same results as Lasso, indicating that it might not be capturing the underlying data variability effectively. This can happen when the model is too regularized, leading to underfitting.
Model Comparison:

When comparing predictions across models, Ridge models produced more varied outputs, while Lasso and ElasticNet resulted in constant predictions. This could suggest that Ridge Regression is more sensitive to the input data variability compared to the Lasso and ElasticNet models, which might be reducing complexity excessively.
Hyperparameter Tuning Impact:

The hyperparameter-tuned models (Ridge HT, Lasso HT, and ElasticNet HT) reflect the same patterns as their respective non-tuned versions. This indicates that hyperparameter tuning did not significantly change the predictive power of these models in this case.
Evaluation Considerations:

The observed predictions raise concerns about model performance and generalization. Consider evaluating the models using metrics such as Mean Squared Error (MSE) and R-squared values to quantify their accuracy and fit.
The model selection may need to be revisited, possibly by exploring additional hyperparameter tuning, feature engineering, or even using ensemble methods for better predictive performance.